### PR TITLE
Dataretriever QueryResponse can show more metadata columns

### LINKED ISSUE #3321: Dataretriever Clients should be able to specify what columns are in their results table
For instance the GOES based satellites probably want satellite number in the table.

This is actually more complex because there isn't a method currently for subclasses of `GenericClient` to return a list of results with meta data for each file.

### PR DESCRIPTION
**Pull request by abhijeetmanhas**
### Description
* `show()` method in `base_client()` to specify which columns to be shown in response.
* Dataretriever clients `QueryResponse` have more metadata info like goes `SatelliteNumber` , `Detector`, `Level`, etc. got from url.
* Correct satellite number returned in goes overlapping urls.

### New metadata cols added for
 * [x] Eve
 * [x] Fermi
 * [x] Lyra
 * [x] NorH
 * [x] XRS
 * [x] SUVI
 * [ ] Rhessi
 * [ ] Noaa
### Updated Tests
 * [x] Docs
 * [ ] Database
 * [x] net
 * [x] util
### Fix `parse` dependency
 * [x] Pypi
 * [x] Conda
### Miscellaneous
 * [ ] What to do with `EndTime` for clients?
 * [ ] Add `docs` for new functionality
 * [ ] Add tests for new metadata cols
 * [ ] Discuss better approaches for returning metadata from scrapper

### Some snippets
```python
from sunpy.net import Fido, attrs as a
res=Fido.search(a.Time('2019/05/25 00:50', '2019/05/25 00:52'),a.Instrument('suvi'))
res

<sunpy.net.fido_factory.UnifiedResponse object at 0x7f6c418c2cd0>
Results from 1 Provider:

6 Results from the SUVIClient:
Level SatelliteNumber        StartTime                EndTime           Wavelength   Source Provider Physobs Instrument
----- --------------- ----------------------- ----------------------- -------------- ------ -------- ------- ----------
    2              16 2019-05-25T00:52:00.000 2019-05-25T00:56:00.000  94.0 Angstrom   GOES     NOAA    flux       SUVI
    2              16 2019-05-25T00:52:00.000 2019-05-25T00:56:00.000 131.0 Angstrom   GOES     NOAA    flux       SUVI
    2              16 2019-05-25T00:52:00.000 2019-05-25T00:56:00.000 171.0 Angstrom   GOES     NOAA    flux       SUVI
    2              16 2019-05-25T00:52:00.000 2019-05-25T00:56:00.000 195.0 Angstrom   GOES     NOAA    flux       SUVI
    2              16 2019-05-25T00:52:00.000 2019-05-25T00:56:00.000 284.0 Angstrom   GOES     NOAA    flux       SUVI
    2              16 2019-05-25T00:52:00.000 2019-05-25T00:56:00.000 304.0 Angstrom   GOES     NOAA    flux       SUVI
```
Satellite numbers, wavelength shown, previosuly the response table was only this:
```python
<sunpy.net.fido_factory.UnifiedResponse object at 0x7f8158cb1dd0>
Results from 1 Provider:

6 Results from the SUVIClient:
     Start Time           End Time      Source Instrument Wavelength
------------------- ------------------- ------ ---------- ----------
2019-05-25 00:50:00 2019-05-25 00:52:00   GOES       SUVI        nan
2019-05-25 00:50:00 2019-05-25 00:52:00   GOES       SUVI        nan
2019-05-25 00:50:00 2019-05-25 00:52:00   GOES       SUVI        nan
2019-05-25 00:50:00 2019-05-25 00:52:00   GOES       SUVI        nan
2019-05-25 00:50:00 2019-05-25 00:52:00   GOES       SUVI        nan
2019-05-25 00:50:00 2019-05-25 00:52:00   GOES       SUVI        nan
```
One more example:
```python
res = Fido.search(a.Time('2015-06-21 00:00', '2015-06-23 23:59'),a.Instrument.gbm, a.Detector.n3,a.Resolution.ctime)
dr_res = res.get_response(0)
dr_res.show('Detector','Resolution', 'Phsyobs')

<Table length=3>
Detector Resolution Physobs
  str2      str5      str4 
-------- ---------- -------
      n3      ctime    flux
      n3      ctime    flux
      n3      ctime    flux
```
Fixes #3321, #556 , #4207. Partially  #3715

### CODE REVIEW

**Comment by abhijeetmanhas** [comment]
I think I should open a new Pr for minor fix for XRS goes issue.

**Comment by nabobalis** [comment]
I think it might be nice to add an example into the fido.rst guide or some other page documenting this new behaviour.

File: sunpy/net/dataretriever/client.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ class QueryResponseBlock:
     Represents url, source along with other information
     """
 
-    def __init__(self, map0, url, time=None):
+    def __init__(self, map0, url, time=None, meta=None):
         """
         Parameters
         ----------
```
Comment (nabobalis) [inline sunpy/net/dataretriever/client.py:31]: I think it might be worth adding documentation for the parameters.

File: sunpy/net/dataretriever/client.py [hunk 2]
```diff
@@ -41,12 +41,18 @@ def __init__(self, map0, url, time=None):
         self.time = TimeRange(map0.get('Time_start'),
                               map0.get('Time_end')) if time is None else time
         self.wave = map0.get('wavelength', np.NaN)
+        if meta is not None:
```
Comment (nabobalis) [inline sunpy/net/dataretriever/client.py:44]: What happens if meta is None?

**Comment by nabobalis** [comment]
Also from the weekly call, it would be nice if the columns had a fixed order for the first 5 or so columns.

File: sunpy/net/dataretriever/client.py (hunk 2 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/client.py:44]: Results remain same as they were earlier, no change. If not None, only then it shows new info.

File: setup.cfg [hunk 3]
```diff
@@ -52,6 +52,7 @@ net =
   python-dateutil
   zeep
   tqdm
+  parse
```
Comment (nabobalis) [inline setup.cfg:55]: I wonder if we don't actually add this as a dependency by copy the file to the extern folder instead.

File: sunpy/net/dataretriever/sources/tests/test_norh.py [hunk 4]
```diff
@@ -34,7 +34,7 @@ def LCClient():
      )
 ])
 def test_get_url_for_time_range(LCClient, timerange, url_start, url_end):
-    urls = LCClient._get_url_for_timerange(timerange, wavelength=17*u.GHz)
+    urls, urlmeta = LCClient._get_url_for_timerange(timerange, wavelength=17*u.GHz)
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/tests/test_norh.py:37]: Is it worth testing ``urlmeta`` in any of these tests?

**Comment by nabobalis** [comment]
> I think I should open a new Pr for minor fix for XRS goes issue.

What XRS goes issue?

**Comment by abhijeetmanhas** [comment]
> > I think I should open a new Pr for minor fix for XRS goes issue.
> 
> What XRS goes issue?

#4207 , I think it would be better to add new pr for it.

**Comment by nabobalis** [comment]
> > > I think I should open a new Pr for minor fix for XRS goes issue.
> > 
> > 
> > What XRS goes issue?
> 
> #4207 , I think it would be better to add new pr for it.

Yes a seperate PR for that.

**Comment by abhijeetmanhas** [comment]
now it fixed #3337 .
Still have to figure out generalization for
- `suvi`
- `norh`
- `rhessi`
- `noaa`

And properly document a lot of API.

File: sunpy/net/dataretriever/sources/eve.py [hunk 5]
```diff
@@ -27,52 +19,24 @@ class EVEClient(GenericClient):
...
-
-    def _get_time_for_url(self, urls):
-        eve = Scraper(BASEURL)
-        times = list()
-        for url in urls:
-            t0 = eve._extractDateURL(url)
-            # hard coded full day as that's the normal.
-            times.append(TimeRange(t0, t0 + TimeDelta(1*u.day)))
-        return times
+    baseurl = (r'http://lasp.colorado.edu/eve/data_access/evewebdata/quicklook/'
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/eve.py:36]: These are now class attributes?

File: sunpy/util/scraper.py [hunk 6]
```diff
@@ -128,6 +153,18 @@ def _URL_followsPattern(self, url):
         """
         Check whether the url provided follows the pattern.
         """
+        if self.matcher is not None:
```
Comment (abhijeetmanhas) [inline sunpy/util/scraper.py:156]: matches using the `**kwargs` of client query. It checks whether the info extracted from `url` have attrs one among all possible attrs.

File: sunpy/net/dataretriever/sources/tests/test_goes_ud.py [hunk 7]
```diff
@@ -39,11 +36,11 @@ def test_get_url_for_time_range(LCClient, timerange, url_start, url_end):
...
 @pytest.mark.parametrize("timerange, url_start, url_end",
                          [(TimeRange('1999/01/10', '1999/01/20'),
-                           'https://umbra.nascom.nasa.gov/goes/fits/1999/go10990110.fits',
+                           'https://umbra.nascom.nasa.gov/goes/fits/1999/go0819990115.fits',
                            'https://umbra.nascom.nasa.gov/goes/fits/1999/go1019990120.fits')])
 def test_get_overlap_urls(LCClient, timerange, url_start, url_end):
-    urls = LCClient._get_url_for_timerange(timerange)
-    assert len(urls) == 9
+    urls, urlmeta = LCClient._get_url_for_timerange(timerange)
+    assert len(urls) == 28
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/tests/test_goes_ud.py:43]: So our old requests get more results now?

File: sunpy/util/scraper.py [hunk 8]
```diff
@@ -90,10 +94,31 @@ def __init__(self, pattern, regex=False, **kwargs):
                 milli=milliseconds_,
                 end=self.pattern[milliseconds.end():]
             ))
+        self.extractor = extractor
+        self.matcher = matcher
 
     def matches(self, filepath, date):
         return date.strftime(self.pattern) == filepath
 
+    def infoextract(self, url, timestamp=False):
```
Comment (abhijeetmanhas) [inline sunpy/util/scraper.py:103]: If timestamp is `true`, we return a url dict containing time info. Else we return all fields except time.

File: sunpy/net/tests/test_fido.py [hunk 9]
```diff
@@ -68,7 +68,7 @@ def test_offline_fido(query):
 # Until we get more mocked, we can't really do this to online clients.
 # TODO: Hypothesis this again
 @pytest.mark.parametrize("query", [
-    (a.Instrument.eve & a.Time('2014/7/7', '2014/7/14') & a.Level.zero),
+    (a.Instrument.eve & a.Time('2014/7/7', '2014/7/14') & a.Level('0CS')),
```
Comment (nabobalis) [inline sunpy/net/tests/test_fido.py:71]:
```suggestion
    (a.Instrument.eve & a.Time('2014/7/7', '2014/7/14') & a.Level('0CS')),
```
How has `0CS` been processed by the attrs registration code.

File: sunpy/util/scraper.py (hunk 8 above)
Comment (nabobalis) [inline sunpy/util/scraper.py:103]: Would need a test.

File: sunpy/util/scraper.py [hunk 10]
```diff
@@ -220,7 +273,7 @@ def filelist(self, timerange):
         >>> solmon = Scraper(solmon_pattern, instrument = 'swap', wave = 174)
         >>> from sunpy.time import TimeRange
         >>> timerange = TimeRange('2015-01-01','2015-01-01T16:00:00')
-        >>> print(solmon.filelist(timerange))  # doctest: +REMOTE_DATA
+        >>> print(solmon.filelist(timerange)[0])  # doctest: +REMOTE_DATA
```
Comment (nabobalis) [inline sunpy/util/scraper.py:276]: Do we get a list of lists?

File: sunpy/util/scraper.py [hunk 11]
```diff
@@ -196,6 +238,17 @@ def url_to_list(txt):
         return Time.strptime(' '.join(final_date),
                              ' '.join(final_pattern))
 
+    def _extractMetaDataURL(self, urls):
```
Comment (nabobalis) [inline sunpy/util/scraper.py:241]: Would need a test?

File: sunpy/net/dataretriever/sources/eve.py (hunk 5 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/eve.py:36]: yes, only a baseurl and an extractor are needed to define most of the clients.

File: sunpy/net/dataretriever/sources/tests/test_eve.py [hunk 12]
```diff
@@ -117,7 +114,7 @@ def test_levels(time):
 
 def test_attr_reg():
     assert a.Instrument.eve == a.Instrument('EVE')
-    assert a.Level.zero == a.Level('0')
+    assert a.Level('0cs') == a.Level('0cs')
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/tests/test_eve.py:117]: This test was to make sure the `'0cs'` key is properly processed by the attrs reg code.

File: sunpy/net/dataretriever/client.py [hunk 13]
```diff
@@ -207,7 +227,42 @@ def _get_url_for_timerange(cls, timerange, **kwargs):
...
+        that can be retrieved from the file url.
+
+        Parameters
+        ----------
+        \\*\\*kwargs: `dict`
+             Any extra keywords to refine the search. Generated from the
+             attributes passed to
+             `~sunpy.net.dataretriever.client.GenericClient.search`.
+        """
+        a = cls.register_values()
```
Comment (nabobalis) [inline sunpy/net/dataretriever/client.py:249]: I wonder how many times we call this and is it worth storing as a dict on the class itself.

File: sunpy/net/dataretriever/sources/tests/test_goes_ud.py (hunk 7 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/tests/test_goes_ud.py:43]: This was because if we don't define specify level in query of `lyra`, data for level 1,2,3 are returned. No level is favoured or defaulted over other.

File: sunpy/net/dataretriever/sources/tests/test_eve.py (hunk 12 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/tests/test_eve.py:117]: It was difficult to get alternatives values for registered attrs. Actually in `register_values`, it can take only one possible value, `0` or `0cs`. So I fixed it ``0cs

Comment (nabobalis) [reply sunpy/net/dataretriever/sources/tests/test_eve.py:117]: So is there a `a.Level.0cs`?

File: sunpy/net/dataretriever/sources/noaa.py [hunk 14]
```diff
@@ -26,15 +29,15 @@ class NOAAIndicesClient(GenericClient):
 
     >>> from sunpy.net import Fido, attrs as a
     >>> results = Fido.search(a.Time("2016/1/1", "2016/1/2"),
-    ...                       a.Instrument.noaa_indices)  #doctest: +REMOTE_DATA
+    ...                       a.Instrument("noaa-indices"))  #doctest: +REMOTE_DATA
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/noaa.py:32]: Why was this changed?

File: sunpy/net/dataretriever/sources/noaa.py [hunk 15]
```diff
@@ -102,23 +105,19 @@ class NOAAPredictClient(GenericClient):
...
+    Source Provider    Physobs      Instrument
+    ------ -------- -------------- ------------
+      ises     swpc sunspot number noaa-predict
     <BLANKLINE>
     <BLANKLINE>
 
     """
-    @staticmethod
-    def _get_default_uri():
-        """Return the url to download indices"""
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/noaa.py:114]: This comment should be moved to where the URL is set.

File: sunpy/net/dataretriever/sources/noaa.py [hunk 16]
```diff
@@ -102,23 +105,19 @@ class NOAAPredictClient(GenericClient):
...
-        return ["http://services.swpc.noaa.gov/text/predicted-sunspot-radio-flux.txt"]
 
     def _get_url_for_timerange(self, timerange, **kwargs):
         """
         Helper function:
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/noaa.py:118]: This either needs updating or removing.

File: setup.cfg [hunk 17]
```diff
@@ -51,6 +51,7 @@ net =
...
```
Comment (nabobalis) [inline setup.cfg:54]: This can be removed now.

**Comment by abhijeetmanhas** [comment]
Upgraded to #4321
